src/sync.jl

```diff
@@ -0,0 +1,65 @@
+module Sync
+    global mutex = ReentrantLock()
+    global sync_level = repeat([ 0 ], Threads.nthreads())
+
+    @inline is_locked() = sync_level[Threads.threadid()] > 0
+
+    @inline function lock()
+        tid = Threads.threadid()
+        if sync_level[tid] == 0
+            Base.lock(mutex)
+        end
+        sync_level[tid] += 1
+    end
```

Comment:
I don't understand: We are using a `ReentrantLock`. Why do we need to keep track on every thread whether we already locked the lock? I.e. why do we need `sync_level`?

Is it because locks are per-task but you want per-thread locks? But why? Shouldn't we prevent any attempts to invoke GAP simultaneously from two different tasks, even if they happen to run on the same thread? I mean, GAP still won't expect two concurrent executions and things can and will break in various ways, no?

(BTW, note that at least in some build configurations, tasks can migrate between threads... however, I have no idea if this can happen while the task holds a lock is being held... hopefully not).

Reply:
This is something that's still a bit of a work in progress.

Julia has `ReentrantLock`, which is reentrant on a per-task basis. It also has `SpinLock`, which works on a per-thread basis, but will eat CPU while blocking. So, in a way, that's a placeholder implementation, and I think I may actually have to wrap a pthread mutex for this to work properly.

Even if `ReentrantLock` works, the `lock()` method for `ReentrantLock()` is comparatively expensive, which is why I tried to optimize it for the nested case.

(`ReentrantLock` working on a per-task basis is also what I suspect may be one of the causes for the crashes I'm seeing.)